lxd/storage/utils.go

```diff
@@ -0,0 +1,406 @@
+	case "btrfs":
+		msg, err = shared.TryRunCommand("btrfs", "filesystem", "resize", "max", mntpoint)
+	default:
+		return fmt.Errorf(`Growing not supported for filesystem type "%s"`, fsType)
+	}
+
+	if err != nil {
+		errorMsg := fmt.Sprintf(`Could not extend underlying %s filesystem for "%s": %s`, fsType, devPath, msg)
+		logger.Errorf(errorMsg)
+		return fmt.Errorf(errorMsg)
+	}
+
+	logger.Debugf(`extended underlying %s filesystem for "%s"`, fsType, devPath)
+	return nil
+}
+
+func shrinkFileSystem(fsType string, devPath string, mntpoint string, byteSize int64) error {
+	strSize := fmt.Sprintf("%dK", byteSize/1024)
+
+	switch fsType {
+	case "": // if not specified, default to ext4
+		fallthrough
+	case "ext4":
+		_, err := shared.TryRunCommand("e2fsck", "-f", "-y", devPath)
+		if err != nil {
+			return err
+		}
+
+		_, err = shared.TryRunCommand("resize2fs", devPath, strSize)
+		if err != nil {
+			return err
+		}
+	case "btrfs":
+		_, err := shared.TryRunCommand("btrfs", "filesystem", "resize", strSize, mntpoint)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf(`Shrinking not supported for filesystem type "%s"`, fsType)
+	}
+
+	return nil
+}
+
+/*
+func shrinkVolumeFilesystem(s StorageDriver, volumeType int, fsType string, devPath string, mntpoint string, byteSize int64, data interface{}) (func() (bool, error), error) {
+	var cleanupFunc func() (bool, error)
+	switch fsType {
+	case "xfs":
+		logger.Errorf("XFS filesystems cannot be shrunk: dump, mkfs, and restore are required")
+		return nil, fmt.Errorf("xfs filesystems cannot be shrunk: dump, mkfs, and restore are required")
+	case "btrfs":
+		fallthrough
+	case "": // if not specified, default to ext4
+		fallthrough
+	case "ext4":
+		switch volumeType {
+		case storagePoolVolumeTypeContainer:
+			c := data.(Container)
+			ourMount, err := c.StorageStop()
+			if err != nil {
+				return nil, err
+			}
+			if !ourMount {
+				cleanupFunc = c.StorageStart
+			}
+		case storagePoolVolumeTypeCustom:
+			ourMount, err := s.StoragePoolVolumeUmount()
+			if err != nil {
+				return nil, err
+			}
+			if !ourMount {
+				cleanupFunc = s.StoragePoolVolumeMount
+			}
+		default:
+			return nil, fmt.Errorf(`Resizing not implemented for storage volume type %d`, volumeType)
+		}
+
+	default:
+		return nil, fmt.Errorf(`Shrinking not supported for filesystem type "%s"`, fsType)
+	}
+
+	err := shrinkFileSystem(fsType, devPath, mntpoint, byteSize)
+	return cleanupFunc, err
+}
+*/
+
+// Returns the parent container name, snapshot name, and whether it actually was
+// a snapshot name.
+func containerGetParentAndSnapshotName(name string) (string, string, bool) {
+	fields := strings.SplitN(name, shared.SnapshotDelimiter, 2)
+	if len(fields) == 1 {
+		return name, "", false
+	}
+
+	return fields[0], fields[1], true
+}
+
+// /var/lib/lxd/[snapshots|containers]/name
+func containerPath(projectName string, name string, isSnapshot bool) string {
```

Comment:
Defined but not used in this commit

Reply:
I created a commit which moves this function to shared.